File: cses/03_dynamic_programming/1653_elevator-rides/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    x = int(data[1])
    w = [int(data[2 + i]) for i in range(n)]

    # Mã hoá cặp (rides, weight_chuyến_cuối) thành một số: state = rides * BIG + weight.
    # BIG = x + 1 vì weight luôn <= x, nên lấy min trên số này chính là so sánh
    # lexicographic (ưu tiên rides nhỏ, rồi tới weight của chuyến cuối).
    BIG = x + 1
    size = 1 << n
    full = size - 1

    dp = [0] * size
    dp[0] = BIG                             # Base case: 1 chuyến mở, chưa chở ai (rides=1, weight=0).

    for mask in range(1, size):
        best = 1 << 62
        m = mask
        # Thử chọn từng người p là người được thêm cuối cùng vào mask.
        while m:
            low = m & (-m)
            p = low.bit_length() - 1
            prev = dp[mask ^ low]
            wp = w[p]
            wt = prev % BIG                 # Weight của chuyến cuối ở state prev.
            if wt + wp <= x:                # Còn chỗ -> nhét p vào chuyến đang mở.
                cand = prev + wp
            else:                           # Chuyến cuối đã đầy -> mở chuyến mới cho p.
                cand = prev - wt + BIG + wp
            if cand < best:
                best = cand
            m ^= low
        dp[mask] = best

    # Đáp án là phần rides của state đầy đủ.
    sys.stdout.write(str(dp[full] // BIG) + "\n")


main()

File: cses/03_dynamic_programming/1653_elevator-rides/solution.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    long long x;
    cin >> n >> x;

    vector<long long> weight(n);
    for (long long& value : weight) {
        cin >> value;
    }

    // Mã hoá cặp (rides, last_weight) thành một số: rides * BIG + last_weight.
    // BIG = x + 1 vì cân nặng chuyến cuối luôn <= x, nên lấy min trên số này
    // chính là so sánh lexicographic (ưu tiên rides nhỏ, rồi tới last_weight).
    const long long BIG = x + 1;
    const int size = 1 << n;
    vector<long long> dp(size);
    dp[0] = BIG;  // Base case: đã có 1 chuyến mở, chưa chở ai, cân nặng 0.

    for (int mask = 1; mask < size; ++mask) {
        long long best = numeric_limits<long long>::max();
        int remaining = mask;
        // Thử chọn từng người person là người được thêm cuối cùng vào mask.
        while (remaining != 0) {
            const int bit = remaining & -remaining;
            const int person = __builtin_ctz(static_cast<unsigned>(bit));
            const long long previous = dp[mask ^ bit];
            const long long last_weight = previous % BIG;

            long long candidate;
            if (last_weight + weight[person] <= x) {
                // Còn chỗ trong chuyến đang mở -> nhét person vào chuyến đó.
                candidate = previous + weight[person];
            } else {
                // Chuyến cuối đã đầy -> mở chuyến mới chỉ chứa person.
                candidate = previous - last_weight + BIG + weight[person];
            }
            best = min(best, candidate);
            remaining ^= bit;
        }
        dp[mask] = best;
    }

    // Đáp án là phần rides của state đầy đủ.
    cout << dp[size - 1] / BIG << '\n';
    return 0;
}
